proposals/0000-import-renaming.md

```diff
@@ -0,0 +1,141 @@
+---
+author: Oleg Grenrus
+date-accepted: ""
+ticket-url: ""
+implemented: ""
+---
+
+This proposal is [discussed at pull request 408](https://github.com/ghc-proposals/ghc-proposals/pull/408)
+
+# Import renaming
+
+We propose to add `-XRenamingImports` to extend import syntax such that it is possible to rename definitions.
+
+```haskell
+import Distribution.Simple renaming (defaultMain to main)
+```
+
+## Motivation
+
+Names are source of many debates. Whether
+
+```haskell
+newtype Fix = Fix { unFix :: f (Fix a) }
+```
+
+or
+
+```haskell
+newtype Fix = Fix { unfix :: f (Fix a) }
+```
+
+Recently implemented `QualifiedDo` extension requires to create dummy modules just for this feature.
+Instead we could
+
+```haskell
+import qualified Data.Functor.Bind as Semi renaming ((>>-) to (>>=))
+```
+
+Another example is `optics`. Some people just like `.` as optical composition operator.
+
+```haskell
+import Prelude hiding ((.))
+import Optics renaming ((%) to (.))
+```
+
+This is a small step towards [*Local modules* discussed in PR#283](https://github.com/ghc-proposals/ghc-proposals/pull/283).
+The proposals are not in conflict.
+
+## Proposed Change Specification
+
+```
+impdecl	→	import [qualified] modid [as modid] [impspec] [renspec]
+
+renspec →   renaming ( import1 to rename1 , … , importn to rename2 [ , ] )     (n ≥ 0)
+
+rename  →   almost like import, but without namespaces
+```
+
+
+Lexically `renaming` and `to` are `varid` (similarly to `as`, `qualified` and `hiding`) rather than a `reservedid`.
+They have special significance only in the context of `import` declaration, it may also be used as a variable.
+
+There is no conflict with `import Foo qualified` syntax.
+
+
+## Examples
+
+We can hide an rename.
+
+```haskell
+import Prelude hiding (tail) renaming (head to unsafeHead)
+```
+
+We can import just some names, and rename others
+
+```haskell
+import Data.List (foldl') renaming (foldr1 to unsafeFoldr1)
+```
+
+Renamed names are always imported, `importspec` specifies
+further names to be imported or hidden.
+
+We can rename with `ExplicitNamespaces`
+
+```haskell
+import GHC.Generics renaming (type (:*:) to Prod)
+```
+
+We can rename import declaration
+
+```haskell
+import Data.Maybe renaming (Maybe (Just, Nothing) to Option (Some, None))
+```
+
+## Effect and Interactions
+
+TBW
+
+## Costs and Drawbacks
+
+Allowing to *export* renamed names would allow a lot,
+but would require GHC interface files to contain
+not only bindings but also renamings.
+
+There is interesting interaction with `RecordDotSyntax`-like features,
+where constraint solver generates instances based on names in scope.
+Yet it wouldn't be possible to rename instances defined by hand.
+(TODO Check: if it's forbidden to shadow automatic instance, the renaming will break that invariant).
```

Comment:
Indeed, at the moment hand-written `HasField` instances are required not to overlap with existing fields of the type. This proposal would allow you to introduce use sites where a hand-written `HasField` instance overlaps with the built-in constraint solver behaviour.

I don't think that's a huge problem though: we can just reject such programs as if the constraint solver got stuck due to two overlapping instances. The only dark corner I can see is that if you define a record pattern synonym and an accompanying manual `HasField` instance, renaming would affect the pattern synonym field but not the `HasField` instance.

```hs
module M where
  data T = MkT { foo :: Int }

  instance HasField "bar" T Bool where
    getField _ = True

module N where
  import M renaming (foo to bar)

  hmm = getField @"bar"
```

Reply:
@adamgundry how record patterns and `HasField` interact? It's possible to have both now as well? To answer this questio myself: in GHC-9.0 pattern synonyms don't affect `HasField` resolution. It can be argued it should or shouldn't. Should because pattern synonyms are supposed to be indistinguishable from real constructors/fields. shouldn't because things would break & not compose :)

## An example

### A.hs

```haskell
module A where

data Record = Record { recFieldA :: Int, recFieldB :: Bool }
  deriving Show
```

### B.hs

Introduces pseudo fields using `HasField`

```haskell
{-# LANGUAGE DataKinds, MultiParamTypeClasses #-}
module B where

import A
import GHC.Records

instance HasField "recFieldC" Record Int where
    getField = recFieldA
``` 

### C.hs

Introduces pseudo fields using `PatternSynonyms`

```haskell
{-# LANGUAGE PatternSynonyms #-}
module C where

import A

pattern Pattern :: Int -> Bool -> Record
pattern Pattern{recFieldC,recFieldD} = Record recFieldC recFieldD
```

### D.hs

Use site

```haskell
{-# LANGUAGE DataKinds, TypeApplications #-}
module D where

import A
import GHC.Records

import B
import C

record :: Record
record = Record 42 True

example :: Int
example = getField @"recFieldC" record
```

This works because a solver for `HasField` doesn't look for pattern synonyms.

---

This (`ImportRenaming`) proposal would introduce yet another way
to do that

I.e.

```haskell
import A renaming (recFieldA to recFieldC)
```

I guess that **taking renaming into accounts** when solving for
`HasField` would be easier implementation wise,
and less surprising for users,
but it's also possible to consider only real names.

---

**Important observation** (not related to this proposa): With `RecordDotSyntax` which builds on top
of `HasField` the renaming the original fields is a breaking change,
when backward-compatibility is provided by `PatternSynonyms`.

Reply:
Yes, I agree with all this. And indeed it is a bit of an awkward point that using a record pattern synonym for backwards compatibility can be broken by `HasField`; while working around this with a manual instance is possible, it's hardly pleasant. 

The underlying reason why pattern synonyms don't lead to `HasField` constraints being solved is that they can insert "fields" into arbitrary types (or even be completely polymorphic). Thus they may conflict with the existing solver behaviour or with manual `HasField` instances. But perhaps we should treat all these cases more like overlapping instances at use sites, rather than ruling them out too early.